File: src/openmp/plugin-SiPixelRecHits/SiPixelRecHitCUDA.cc
#include "CUDADataFormats/SiPixelClustersSoA.h"
#include "CUDADataFormats/SiPixelDigisSoA.h"
#include "CUDADataFormats/TrackingRecHit2DHeterogeneous.h"
#include "DataFormats/BeamSpotPOD.h"
#include "Framework/EventSetup.h"
#include "Framework/Event.h"
#include "Framework/PluginFactory.h"
#include "Framework/EDProducer.h"
#include "CondFormats/PixelCPEFast.h"

#include "PixelRecHits.h"  // TODO : spit product from kernel

class SiPixelRecHitCUDA : public edm::EDProducer {
public:
  explicit SiPixelRecHitCUDA(edm::ProductRegistry& reg);
  ~SiPixelRecHitCUDA() override = default;

private:
  void produce(edm::Event& iEvent, const edm::EventSetup& iSetup) override;

  // The mess with inputs will be cleaned up when migrating to the new framework
  edm::EDGetTokenT<BeamSpotPOD> tBeamSpot;
  edm::EDGetTokenT<SiPixelClustersSoA> token_;
  edm::EDGetTokenT<SiPixelDigisSoA> tokenDigi_;

  edm::EDPutTokenT<TrackingRecHit2DCPU> tokenHit_;

  pixelgpudetails::PixelRecHitGPUKernel gpuAlgo_;
};

SiPixelRecHitCUDA::SiPixelRecHitCUDA(edm::ProductRegistry& reg)
    : tBeamSpot(reg.consumes<BeamSpotPOD>()),
      token_(reg.consumes<SiPixelClustersSoA>()),
      tokenDigi_(reg.consumes<SiPixelDigisSoA>()),
      tokenHit_(reg.produces<TrackingRecHit2DCPU>()) {}

void SiPixelRecHitCUDA::produce(edm::Event& iEvent, const edm::EventSetup& es) {
  PixelCPEFast const& fcpe = es.get<PixelCPEFast>();

  auto const& clusters = iEvent.get(token_);
  auto const& digis = iEvent.get(tokenDigi_);
  auto const& bs = iEvent.get(tBeamSpot);

  auto nHits = clusters.nClusters();
  if (nHits >= TrackingRecHit2DSOAView::maxHits()) {
    std::cout << "Clusters/Hits Overflow " << nHits << " >= " << TrackingRecHit2DSOAView::maxHits() << std::endl;
  }

  iEvent.emplace(tokenHit_, gpuAlgo_.makeHits(digis, clusters, bs, &fcpe.getCPUProduct()));
}

DEFINE_FWK_MODULE(SiPixelRecHitCUDA);

File: src/julia-serial/plugin-SiPixelRecHits/SiPixelRecHitCUDA.jl
using .CUDADataFormatsSiPixelDigiInterfaceSiPixelDigisSoA
using .CUDADataFormatsSiPixelClusterInterfaceSiPixelClustersSoA
using .CUDADataFormats_TrackingRecHit_interface_TrackingRecHit2DHeterogeneous_h
using .CUDADataFormats_TrackingRecHit_interface_TrackingRecHit2DSOAView_h
using .RecoLocalTracker_SiPixelRecHits_plugins_PixelRecHits_h
using .PluginFactory

# more includes are missing (waiting on files to be done)

struct SiPixelRecHitCUDA <: EDProducer
    tBeamSpot::EDGetTokenT{BeamSpotPOD}
    token::EDGetTokenT{SiPixelClustersSoA}
    tokenDigi::EDGetTokenT{SiPixelDigisSoA}
    tokenHit::EDPutTokenT{TrackingRecHit2DHeterogeneous}

    function SiPixelRecHitCUDA(reg::ProductRegistry)
        tBeamSpot= consumes(reg, BeamSpotPOD)
        token = consumes(reg, SiPixelClustersSoA)
        tokenDigi = consumes(reg, SiPixelDigisSoA)
        tokenHit = produces(reg, TrackingRecHit2DHeterogeneous)
        new(tBeamSpot, token, tokenDigi, tokenHit)
    end
end

function produce(self::SiPixelRecHitCUDA,iEvent::Event, es::EventSetup)
    fcpe = get(es, PixelCPEFast)
    
    clusters = get(iEvent, self.token)
    
    digis = get(iEvent, self.tokenDigi)
    
    bs = get(iEvent, self.tBeamSpot)
    
    nHits = nClusters(clusters)
    # println(nHits, "id: $(iEvent.eventId)")
    if nHits >= max_hits()
        println("Clusters/Hits Overflow ",nHits," >= ", TrackingRecHit2DSOAView::maxHits())
    end
    #makeHits(digis, clusters, bs, getCPUProduct(fcpe))
    emplace(iEvent, self.tokenHit, makeHits(digis, clusters, bs, getCPUProduct(fcpe)))
    
end


add_plugin_module("SiPixelRecHitCUDA",x -> SiPixelRecHitCUDA(x))
